test/extensions/stats_sinks/hystrix/hystrix_test.cc

```diff
@@ -0,0 +1,348 @@
+    ON_CALL(cluster1_.success_counter_, value()).WillByDefault(Return((i + 1) * success_step));
+    cluster1_.cluster_info_->stats().upstream_rq_timeout_.add(timeout_step);
+    cluster1_.cluster_info_->stats().upstream_rq_pending_overflow_.add(rejected_step);
+    sink_->flush(source_);
+  }
+
+  std::string rolling_map = sink_->printRollingWindows();
+  EXPECT_NE(std::string::npos, rolling_map.find(cluster1_name_ + ".total"));
+
+  // Check stream format and data.
+  validate_results(TestUtility::bufferToString(buffer), success_step, error_step, timeout_step, 0,
+                   rejected_step, window_size_);
+
+  // Check the values are reset.
+  buffer.drain(buffer.length());
+  sink_->resetRollingWindow();
+  sink_->flush(source_);
+  validate_results(TestUtility::bufferToString(buffer), 0, 0, 0, 0, 0, window_size_);
+}
+
+//
+TEST_F(HystrixSinkTest, Disconnect) {
+  InSequence s;
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+
+  sink_->flush(source_);
+  EXPECT_EQ(buffer.length(), 0);
+
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+  sink_->flush(source_);
+  EXPECT_EQ(getStreamField(TestUtility::bufferToString(buffer), "rollingCountSuccess"), "0");
+  EXPECT_NE(buffer.length(), 0);
+
+  // Disconnect.
+  buffer.drain(buffer.length());
+  sink_->unregisterConnection(&callbacks_);
+  sink_->flush(source_);
+  EXPECT_EQ(buffer.length(), 0);
+
+  // Reconnect.
+  buffer.drain(buffer.length());
+  sink_->registerConnection(&callbacks_);
+  sink_->flush(source_);
+  EXPECT_EQ(getStreamField(TestUtility::bufferToString(buffer), "rollingCountSuccess"), "0");
+  EXPECT_NE(buffer.length(), 0);
+}
+
+TEST_F(HystrixSinkTest, AddAndRemoveClusters) {
+  InSequence s;
+
+  // Arbitrary numbers for testing. Make sure error > timeout.
+  uint64_t success_step = 7;
+  uint64_t error_step = 17;
+  uint64_t timeout_step = 1;
+  uint64_t timeout_retry_step = 2;
+  uint64_t rejected_step = 8;
+
+  uint64_t success_step2 = 3;
+  uint64_t error_4xx_step2 = 1;
+  uint64_t error_4xx_retry_step2 = 2;
+  uint64_t error_5xx_step2 = 3;
+  uint64_t error_5xx_retry_step2 = 4;
+  uint64_t timeout_step2 = 3;
+  uint64_t timeout_retry_step2 = 1;
+  uint64_t rejected_step2 = 5;
+
+  // Register callback to sink.
+  sink_->registerConnection(&callbacks_);
+  Buffer::OwnedImpl buffer = createClusterAndCallbacks();
+
+  // Add new cluster.
+  const std::string cluster2_name{"test_cluster2"};
+  ClusterTestInfo cluster2(cluster2_name);
+  cluster_map_.emplace(cluster2_name, cluster2.cluster_);
+  // Redefining since cluster_map_ is returned by value.
+  ON_CALL(cluster_manager_, clusters()).WillByDefault(Return(cluster_map_));
+
+  // Generate data to both clusters.
+  sink_->flush(source_);
+  for (uint64_t i = 0; i < (window_size_ + 1); i++) {
+    buffer.drain(buffer.length());
+    // Cluster 1
+    ON_CALL(cluster1_.error_5xx_counter_, value()).WillByDefault(Return((i + 1) * 17));
```

Comment:
is the 17 here the same thing as error_step?  If so, use the constant?

Reply:
Right, forgot to use all those values after I defined them...